### PR TITLE
Implemented distance from ball invariant

### LINKED ISSUE #1020: Distance from ball invariant
Maybe multiple invariants, close medium and far?

### PR DESCRIPTION
**Pull request by timovdk**
Returns 0-255 for the closest robot to a ball. The values 0-255 relate to a difference in distance of 0-2 meters (so fuzzy value 127 is 1 meter distance from the ball). This relation between fuzzy value and distance still needs to be tweaked.
closes #1020

### CODE REVIEW

File: src/stp/invariants/DistanceFromBallInvariant.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+     * Creates a piecewise linear function that looks as follows:
+     *
+     * (0,255)  |XX
+     *          | XXXXXX
+     *          |      XXXXXX
+     *          |           XXXXXX
+     *   (0,0)  |----------------XXXXXXX
+     *              (Distance to Ball)
+     */
+    piecewiseLinearFunction = nativeformat::param::createParam(0, 255, 0, "testParam");
```
Comment (jordiluong) [inline src/stp/invariants/DistanceFromBallInvariant.cpp]: testParam: Is this a good name?

Comment (jordiluong) [inline src/stp/invariants/DistanceFromBallInvariant.cpp]: Maybe use FUZZY_TRUE and FUZZY_FALSE for 0 and 255

**Review by jordiluong — APPROVED** [review]

File: src/stp/invariants/DistanceFromBallInvariant.cpp (hunk 1 above)
Comment (timovdk) [reply src/stp/invariants/DistanceFromBallInvariant.cpp]: Woops, missed that

Comment (timovdk) [reply src/stp/invariants/DistanceFromBallInvariant.cpp]: Yes

File: src/stp/invariants/DistanceFromBallInvariant.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    piecewiseLinearFunction->setYAtX(control_constants::FUZZY_TRUE, 0.0);
+    piecewiseLinearFunction->linearRampToYAtX(control_constants::FUZZY_FALSE, stp::control_constants::BALL_IS_CLOSE * 4 + stp::control_constants::FUZZY_MARGIN);
+    piecewiseLinearFunction->setYAtX(control_constants::FUZZY_FALSE, stp::control_constants::BALL_IS_CLOSE * 4 + stp::control_constants::FUZZY_MARGIN);
+}
+
+uint8_t DistanceFromBallInvariant::metricCheck(world_new::view::WorldDataView world, const world::Field* field) const noexcept {
+    auto& us = world.getUs();
+    auto ballPos = world.getBall()->get()->getPos();
+    std::vector<double> distanceMetrics{};
+    distanceMetrics.reserve(control_constants::MAX_ROBOT_COUNT);
```
Comment (RobotJente) [inline src/stp/invariants/DistanceFromBallInvariant.cpp]:
```suggestion
    distances.reserve(control_constants::MAX_ROBOT_COUNT);
```

**Review by RobotJente — APPROVED** [review]

**Pull request merged**